### PR TITLE
Add free-slip, partial-slip and no-slip momentum BCs

### PR DESCRIPTION
**Pull request by cbegeman**
This PR updates the way horizontal momentum boundary conditions are computed --- highlighting that the current MPAS discretisation applies no-slip conditions, introducing new free-slip and partial-slip conditions, and cleaning-up the way cell-vertex remapping is handled across partially masked cells/duals in general.

A new `lateral_walls` namelist section for BC settings is introduced, where `config_wall_slip_factor` controls the amount of slip at horizontal walls (both coastlines at the surface and masked boundaries against bathymetry in general). `config_wall_slip_factor = 0.0` is a no-slip condition, `config_wall_slip_factor = 1.0` is a free-slip condition, and values between 0.0 and 1.0 correspond to partial-slip cases.

In the current implementation the thicknesses from the two valid cells is remapped along with a 0 value for the masked cell. The result is divided by the full area of the dual (triangle) resulting in a vertex layer-thickness of only $\frac{2}{3} h$, inconsistent with the spatially uniform interior thickness. An alternative approach implemented here is to accumulate only the unmasked area of the dual (triangle) and use this in the remapping.

[NCC] - free wall slip (new behavior, activated by namelist change)
[NCC] - no wall slip (default behavior in all configurations)

Credit to @dengwirda for discovering and fixing the layerThicknessVertex bug and for this new feature

### CODE REVIEW

File: components/mpas-ocean/bld/namelist_files/namelist_defaults_mpaso.xml [hunk 1]
```diff
@@ -60,6 +60,9 @@
 <config_time_integrator>'split_explicit_ab2'</config_time_integrator>
 <config_number_of_time_levels>2</config_number_of_time_levels>
 
+<!-- lateral_walls -->
+<config_wall_slip_factor>0.0</config_wall_slip_factor>
```
Comment (cbegeman) [inline components/mpas-ocean/bld/namelist_files/namelist_defaults_mpaso.xml]:
```suggestion
<config_wall_slip_factor>1.0</config_wall_slip_factor>
```

File: components/mpas-ocean/src/Registry.xml [hunk 2]
```diff
@@ -221,6 +221,12 @@
 					possible_values="Any integer greater than or equal to 2."
 		/>
 	</nml_record>
+	<nml_record name="lateral_walls" mode="forward">
+		<nml_option name="config_wall_slip_factor" type="real" default_value="0.0"
```
Comment (cbegeman) [inline components/mpas-ocean/src/Registry.xml]:
```suggestion
		<nml_option name="config_wall_slip_factor" type="real" default_value="1.0"
```

**Comment by cbegeman** [comment]
### Testing

- A 50-year v3.LR.piControl run with free-slip boundary conditions was run. A summary of results is here: https://github.com/E3SM-Ocean-Discussion/E3SM/pull/49#issuecomment-3476850658
- Several idealized test cases were run for validation as described here: https://github.com/E3SM-Ocean-Discussion/E3SM/pull/49#issue-1672937070 and https://github.com/E3SM-Ocean-Discussion/E3SM/pull/49#issuecomment-3275412325
